### PR TITLE
[imex-opt] port serializespirvpass from main branch

### PR DESCRIPTION
**Pull request by Dewei-Wang-sh**
Porting the serializespirv pass from main branch to refactor branch and put the pass in the newly created transform directory

this pass utilizes the upstream spirv::serialize() function to serialize MLIR SPIR-V module to SPIR-V binary and attaches the binary as a string attribute to the gpuModule. 
this pass works like the upstream SerializeToCubin and SerializeToHsaco, only that the other two passes translate llvm dialect to llvm IR and then translate to ISA binary.

Please review these guidelines to help with the review process:
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, a reproducer, or a reference to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?
- [x] Have you organized your commits logically and ensured each can be built by itself?

### CODE REVIEW

File: lib/Transforms/CMakeLists.txt
Comment (Dewei-Wang-sh) [inline lib/Transforms/CMakeLists.txt:5]: will change to imex_source_dir after sang lk's merge

File: test/Transforms/addf_spirv.mlir
Comment (Hardcode84) [inline test/Transforms/addf_spirv.mlir:1]: * Rename file to `serialize-spirv.mlir`
* This test should be added to `main` first and only after that ported to `refactor`

**Review by Hardcode84 — CHANGES_REQUESTED** [review]

File: test/Transforms/addf_spirv.mlir
Comment (Dewei-Wang-sh) [reply test/Transforms/addf_spirv.mlir:1]: thanks for pointing out, recommitted, pls check

**Review by Hardcode84 — COMMENTED** [review]
LGTM, but please wait for review from some `refactor` connoisseurs

File: include/imex/Transforms/Passes.td
Comment (fschlimb) [inline include/imex/Transforms/Passes.td:22]: Please add a description

File: lib/Transforms/SerializeSPIRV.cpp
Comment (fschlimb) [inline lib/Transforms/SerializeSPIRV.cpp:9]: \file

File: lib/Transforms/PassDetail.h
Comment (fschlimb) [inline lib/Transforms/PassDetail.h:8]: \file + description

File: lib/Transforms/SerializeSPIRV.cpp
Comment (fschlimb) [inline lib/Transforms/SerializeSPIRV.cpp:12]: \file + description

File: lib/Transforms/SerializeSPIRV.cpp
Comment (fschlimb) [inline lib/Transforms/SerializeSPIRV.cpp:24]: Could you please add a high-level description of how this is working? There is not a single comment in the code.

File: include/imex/Transforms/Passes.h
Comment (fschlimb) [inline include/imex/Transforms/Passes.h:31]: #undef GEN_PASS_REGISTRATION

File: lib/Transforms/PassDetail.h
Comment (fschlimb) [inline lib/Transforms/PassDetail.h:34]: #undef GEN_PASS_CLASSES

**Review by fschlimb — CHANGES_REQUESTED** [review]
Existing files in refactor use a slightly different header (copyright etc). I suggest using the same everywhere.

File: include/imex/Transforms/Passes.h
Comment (Dewei-Wang-sh) [reply include/imex/Transforms/Passes.h:31]: this undef already exists in the *.inc file, so maybe not needed here

File: lib/Transforms/PassDetail.h
Comment (Dewei-Wang-sh) [reply lib/Transforms/PassDetail.h:34]: this undef already exists in the *.inc file, so maybe not needed here

File: lib/Transforms/SerializeSPIRV.cpp
Comment (Dewei-Wang-sh) [reply lib/Transforms/SerializeSPIRV.cpp:24]: added

File: lib/Transforms/SerializeSPIRV.cpp
Comment (Dewei-Wang-sh) [reply lib/Transforms/SerializeSPIRV.cpp:12]: sure

File: include/imex/Transforms/Passes.td
Comment (Dewei-Wang-sh) [reply include/imex/Transforms/Passes.td:22]: added

**Comment by Dewei-Wang-sh** [comment]
> Existing files in refactor use a slightly different header (copyright etc). I suggest using the same everywhere.

thanks for the detailed review, changes made in a second commit, pls check

File: test/Transforms/addf_spirv.mlir
Comment (nbpatel) [reply test/Transforms/addf_spirv.mlir]: Things don't have to go through main to be added to refactor

**Review by silee2 — APPROVED** [review]
LGTM

File: lib/Transforms/PassDetail.h
Comment (fschlimb) [reply lib/Transforms/PassDetail.h:9]: why did you resolve this without adding the file description?

File: lib/Transforms/SerializeSPIRV.cpp
Comment (Dewei-Wang-sh) [reply lib/Transforms/SerializeSPIRV.cpp]: this view is outdated,  we can see the change in the "file changed" like below
//===- SerializeSPIRV.cpp - SPIR-V serialize pass --------------*- C++ -*-===//
//
// Copyright 2022 Intel Corporation
// Part of the IMEX Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
///
/// \file
/// This pass iterates all the SPIR-V modules in the top module and serializes
/// each SPIR-V module to SPIR-V binary and then attachs the binary blob as a
/// string attribute to the corresponding gpu module.
///
//===---------------

Comment (fschlimb) [reply lib/Transforms/SerializeSPIRV.cpp]: Ok, my bad. Thanks!

**Pull request merged**
